synchronization/src/main/java/de/cyface/synchronization/SyncPerformer.java

```diff
@@ -0,0 +1,125 @@
+import java.net.HttpURLConnection;
+import java.net.ProtocolException;
+import java.net.URL;
+
+/**
+ * Performs the actual synchronisation with a provided server, by uploading meta data and a file containing
+ * measurements.
+ * 
+ * @author Klemens Muthmann
+ * @version 1.0.0
+ * @since 2.0.0
+ */
+class SyncPerformer {
+
+    /**
+     * Triggers the data transmission to a Movebis server API. The <code>measurementIdentifier</code> and
+     * <code>deviceIdentifier</code> need to be globally unique. If they are not the server will probably reject the
+     * request.
+     * <p>
+     * Since this is a synchronous call it can take from seconds to minutes depending on the size of <code>data</code>.
+     * Never call this on the UI thread. Your users are going to hate you.
+     *
+     * @param dataServerUrl The server URL to send the data to.
+     * @param measurementIdentifier The measurement identifier of the transmitted measurement.
+     * @param deviceIdentifier The device identifier of the device transmitting the measurement.
+     * @param data The data to transmit as <code>InputStream</code>.
+     * @param progressListener A listener that is informed about the progress of the upload.
+     * @return The <a href="https://de.wikipedia.org/wiki/HTTP-Statuscode">HTTP status</a> code of the response.
+     */
+    int sendData(final String dataServerUrl, final long measurementIdentifier, final String deviceIdentifier,
+            final @NonNull InputStream data, final @NonNull UploadProgressListener progressListener) {
+        HttpURLConnection.setFollowRedirects(false);
+        HttpURLConnection connection = null;
+        String fileName = String.format("%s_%d.cyf", deviceIdentifier, measurementIdentifier);
+
+        try {
+            connection = (HttpURLConnection)new URL(String.format("%s/measurements", dataServerUrl)).openConnection();
+            try {
+                connection.setRequestMethod("POST");
+            } catch (ProtocolException e) {
+                throw new IllegalStateException(e);
+            }
+            String boundary = "---------------------------boundary";
+            String tail = "\r\n--" + boundary + "--\r\n";
+            connection.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + boundary);
+            connection.setDoOutput(true);
+
+            String userIdPart = addPart("userId", deviceIdentifier, boundary);
+            String measurementIdPart = addPart("measurementId", Long.valueOf(measurementIdentifier).toString(),
+                    boundary);
+
+            String fileHeader1 = "--" + boundary + "\r\n"
+                    + "Content-Disposition: form-data; name=\"fileToUpload\"; filename=\"" + fileName + "\"\r\n"
+                    + "Content-Type: application/octet-stream\r\n" + "Content-Transfer-Encoding: binary\r\n";
+
+            // FIXME This will only work correctly as long as we are using a ByteArrayInputStream. For other streams it
```

Comment:
(Not sure if this is related:) I had troubles using ByteBuffer in the beginning as it's pointer is not always on the first entry ... In the end I figured it out, reading the documentation, see e.g. backend2/GpsPointSerializer, but I can't remember if the problem only was the mixture of read and write access on the Buffer without using .flip().

Reply:
Works so far. There are test cases and as long as they do not fail I have no choice but assuming this code is ok. There will be a few extended tests prior to the campaign, that should uncover any error conditions.